### ISSUE TITLE
scToppR

### ISSUE THREAD

**Issue by bryanwgranger** [Bioconductor/Contributions#3523]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/bryanwgranger/scToppR

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand Bioconductor [Package Naming Policy][9] and acknowledge
  Bioconductor may retain use of package name.

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.
  
- [x] I am familiar with the [Bioconductor code of conduct][7] and 
  agree to abide by it.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For questions/help about the submission process, including questions about
the output of the automatic reports generated by the SPB (Single Package
Builder), please use the #package-submission channel of our Community Slack.
Follow the link on the home page of the [Bioconductor website][8] to sign up.

[1]: https://contributions.bioconductor.org/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/
[7]: https://bioconductor.org/about/code-of-conduct/
[8]: https://bioconductor.org/
[9]: https://bioconductor.org/developers/package-submission/#naming

**Comment by lshep** [comment]
I'm concerned that the usage of the ToppGene website is not better advertised or explained within the package or package documentation.  It should be stated up front in the vignette, man page, and any function that is collecting data from the site that the data might actually be usable and that you would need to check with the "third party"  term and conditions.  

Currently the package vignettes are focused on using data from Seurat Data objects that are not currently common Bioconductor workflow.  It is fine to keep these but it will also be a requirement to show how this can work with and utilize a common Bioconductor class structure for this type of data.

**Comment by lshep** [comment]
may we expect updates soon?

**Comment by bryanwgranger** [comment]
Yes, apologies! Thank you for the feedback, I am finishing up the changes to the code and will be able to commit the changes in the next day or two. When I do, I will explain more here as well.

**Comment by lshep** [comment]
Hopefully there will be updates soon?

**Comment by bryanwgranger** [comment]
Hello, I have made some updates to scToppR:

That's a good point about making sure it's very clear that the data being used is from ToppGene and should adhere to their terms of use. Accordingly, I've added text in the README, a package startup message on attaching, the man file for toppFun(), and the toppFun() function (by printing a message when running it) itself to remind users that all data being used is governed by ToppGene's terms of use.

Additionally, I have also added a new vignette using a standard Bioconductor class, using DESeq2 to analyze the 'airway' dataset. From there, the toppFun() function can be run directly on the results, which are in a DESeqResults object. 

I have pushed all changes to the [Github repository here](https://github.com/bryanwgranger/scToppR). I'm not sure if I need to be pushing the changes at this stage during review to the git.bioconductor.org repo (if that exists), and I'm having trouble doing so. But if I need to, I will address that asap. 

Thanks again for your feedback, please let me know if these changes address your concerns.

**Comment by lshep** [comment]
I'm trying to build this locally and get the following error: 

```
R CMD build scToppR 
* checking for file 'scToppR/DESCRIPTION' ... OK
* preparing 'scToppR':
* checking DESCRIPTION meta-information ... OK
* installing the package to build vignettes
* creating vignettes ... ERROR
--- re-building ‘differential_expression_airway.Rmd’ using rmarkdown

Quitting from lines 24-50 [setup] (differential_expression_airway.Rmd)
Error: processing vignette 'differential_expression_airway.Rmd' failed with diagnostics:
Could not find a root 'DESCRIPTION' file that starts with '^Package' in
'/home/lorikern/biocm/packages/scToppR/'.
ℹ Are you in your project directory and does your project have a 'DESCRIPTION'
  file?
--- failed re-building ‘differential_expression_airway.Rmd’

--- re-building ‘differential_expression_seurat.Rmd’ using rmarkdown

Quitting from lines 25-30 [setup] (differential_expression_seurat.Rmd)
Error: processing vignette 'differential_expression_seurat.Rmd' failed with diagnostics:
Could not find a root 'DESCRIPTION' file that starts with '^Package' in
'/home/lorikern/biocm/packages/scToppR/'.
ℹ Are you in your project directory and does your project have a 'DESCRIPTION'
  file?
--- failed re-building ‘differential_expression_seurat.Rmd’

--- re-building ‘introduction.Rmd’ using rmarkdown

Quitting from lines 28-32 [setup] (introduction.Rmd)
Error: processing vignette 'introduction.Rmd' failed with diagnostics:
Could not find a root 'DESCRIPTION' file that starts with '^Package' in
'/home/lorikern/biocm/packages/scToppR'.
ℹ Are you in your project directory and does your project have a 'DESCRIPTION'
  file?
--- failed re-building ‘introduction.Rmd’

SUMMARY: processing the following files failed:
  ‘differential_expression_airway.Rmd’
  ‘differential_expression_seurat.Rmd’ ‘introduction.Rmd’

Error: Vignette re-building failed.
Execution halted

```

**Comment by bryanwgranger** [comment]
Sorry about that! I've found the error causing that and have updated the vignettes so they should build properly. I tested this in a different environment, and it should work now.

**Comment by lshep** [comment]
I'm not sure how this got through moderation initially as the version number is not Bioconductor standard for incoming packages.  A Bioconductor incoming submission should have a 0.99.0 version number;  have you published this package elsewhere before that requires you to keep 1.0.0?

**Comment by bryanwgranger** [comment]
It is not published elsewhere, that's an oversight. It was originally 0.99.0 and passed moderation that way, but I changed it last week so a colleague could install from github to test it out. I've now changed it back to 0.99.0.

**Comment by bryanwgranger** [comment]
Hello, according to the build report there was 1 WARNING and 1 ERROR. 

-WARNING: "No Bioconductor dependencies detected." scToppR does not require any Bioconductor dependencies, but it can work with standard Bioconductor classes and can be integrated into workflows using other packages easily. I think it will also benefit the Bioconductor community as it is, and there is opportunity to develop more secondary aspects of the package (e.g. plotting using specific Bioconductor packages).

-ERROR:  "Subscribe to the Bioc-devel mailing list by going to https://stat.ethz.ch/mailman/listinfo/bioc-devel" I have now subscribed to the Bioc-devel mailing list.

I am hoping that the warning is ok and the error has been fully resolved and that this can move onto the next review stage.

**Comment by svalvaro** [comment]
Hey @bryanwgranger!
Thanks for the package submission. Before I dig deeper into checking the package, some of the first things I notice are that the package is lacking unit tests. Normally, we use a framework such as [testthat](https://testthat.r-lib.org/) for this. I also noticed that there are no instructions on how to install the package from bioconductor.

**Comment by svalvaro** [comment]
Hi @bryanwgranger could you add instructions for Bioconductor instsallation, also please add some tests to your functions. Probably the `testthat` package is the way to go.

**Comment by bryanwgranger** [comment]
Hey @svalvaro, thank you for the feedback! I have added the instructions for installing from bioconductor. I am also working on unit tests and will have those pushed within a few days.

**Comment by lshep** [comment]
@bryanwgranger  just checking in on further updates. May we expect updates soon?

**Comment by bryanwgranger** [comment]
Hello, I would like to resubmit scToppR - I've updated a few elements that needed it, and I am able to now commit to maintaining the package. Is this possible?

**Comment by lshep** [comment]
Are the git histories consistent that a git push is possible on the version at git.bioconductor.org or would it have to be reset (if divergent histories requiring a force push)

**Comment by bryanwgranger** [comment]
Yes the git histories are consistent, and I've pushed the updated package to git.bioconductor.org.

**Comment by lshep** [comment]
@bryanwgranger could you please push another commit with a version bump so that the package can be rebuilt on our system

**Comment by jwokaty** [comment]
Hi @bryanwgranger ,

Thanks for your patience. I'm using the following checklist to guide the review. Please address the unchecked items all 

# Package Review Checklist {#review-checklist}

**Version 1.0.1**

*scToppR*:
*https://github.com/Bioconductor/Contributions/issues/3523*:

## Package name

- [x] Package name complies the criteria listed on the [Package name section](https://contributions.bioconductor.org/package-name) of the development guide.

## General package development

- [x] `R CMD build` without errors, warnings and notes.
- [  ] Package passes `BiocCheck::BiocCheck()`, `BiocCheck::BiocCheckGitClone()`.
- [x] File names.
- [x] Package size.
- [x] `R CMD check --no-build-vignettes` within 10 minutes.
- [x] Memory requirement below 3GB.
- [x] Size of individual files <= 5MB.
- [x] README file.

## The DESCRIPTION file

Refer to the [DESCRIPTION](https://contributions.bioconductor.org/description.html) section in the BioC package guideline for details on the individual points.

- [x] `Package` field.
- [x] `Title` field.
- [x] `Version` field.
- [x] `Description` field.
- [x] `Authors@R` field.
- [x] `License` field.
- [x] `LazyData` field.
- [x] `Depends`, `Imports`, `Suggests`, `Enhances` fields.
- [x] `SystemRequirements` field.
- [x] `biocViews` field.
- [x] `BugReports` field.
- [x] `URL` field.

## The NAMESPACE file

- [x] Function names use `camelCase` or `snake_case` and do not include `.`.
- [x] Selective imports using `importFrom` instead of *import all* with `import`.
- [x] Individual functions/methods are exported instead of all.

## The NEWS file

- [x] News file in correct location and format. The file should follow versions of scToppR. See https://contributions.bioconductor.org/news.html. I should also be able to get your NEWS via `utils::news`

```
l> library(scToppR)
NOTE: scToppR provides data via ToppGene. Any use of this data must adhere to
  ToppGene's Terms of Use. Please visit https://toppgene.cchmc.org/navigation/termsofuse.jsp
  for more information.
> utils::news(package="scToppR")
NULL
```

## Package data

- [x] Included data not too large. Need for separate data package?
- [x] Exported data and the `data/` directory has correct format, is compressed and documented.

## Documentation

- [x] Vignette present and does describe the core functionality of the package.
- [x] No disabled code blocks present in vignette.
- [x] Vignette uses `BiocStyle` package for formatting.
- [x] Vignette has an *Introduction* section. This should include motivation for inclusion to Bioconductor
- [x] Vignette has an *Installation* section.
- [x] Vignette has a table of contents.
- [x] Vignette includes `sessionInfo()`.
- [x] The `vignettes/` directory contains only vignette file(s) and necessary static images.
- [x] All exported functions and classes have a man page.
- [x] Package man page present--i.e., `?scToppR` displays package man page.
- [x] All man pages have runnable examples.

## Unit tests

- [x] Unit tests present and covering large part of core functionality.

## R code

- [x] All included code under open source license.
- [x] No warnings or errors in `R CMD check`.
- [ ] No warnings or errors in `BiocCheck()`.
- [ ] Coding and syntax (see https://contributions.bioconductor.org/r-code.html):
  - `<-` instead of `=`.
  - `message()`, `warning`, `stop` instead of `cat`. No `paste0` in these
    methods.
  - prefix unexported function names with '.'
- [x] Re-use of classes and functionality (if appropriate).
- [x] Functional programming: no code repetition.
- [x] No excessively long functions.
     * One suggestion regarding code like the following in toppBalloon
      toppBalloon <- function (toppData,
                             categories = NULL,
                             balloons = 3,
                             x_axis_text_size = 6,
                             cluster_col="Cluster",
                             filename = NULL,
                             save = FALSE,
                             height = 5,
                             width = 10) {
                             ...
      if (is.null(filename)) {
        plot_filename <- stringr::str_glue("toppBalloon_{cat}.pdf")
      } else {
        plot_filename = stringr::str_glue("{filename}_{cat}.pdf")
      }
    why not set the default for filename to "toppBalloon" instead of NULL? And move the check up earlier with a `stop` 
- [x] Function argument names descriptive and documented.
- [x] Function arguments should have defaults.
- [x] Function arguments are tested for validity.
- [x] Vectorize: no unnecessary `for` loops present.
- [x] Downloaded files cached with `BiocFileCache`.
- [x] Additional files and dependencies: nothing installed on a user's system.

## Miscellaneous

- [x] Correct Installation section of README: Bioconductor should be Bioconductor, `install.packages('BiocManager')` should be indented.
- [x] Add data provenance to the man pages for pmbc.markers and toppData
- In the Introduction to scToppR for `toppBalloon`, I see a lot of warnings:
  ```toppBalloon(toppData, categories = "GeneOntologyMolecularFunction")
  #> Balloon Plot: GeneOntologyMolecularFunction
  #> Warning in mean.default(sort(x, partial = half + 0L:1L)[half + 0L:1L]):
  #> argument is not numeric or logical: returning NA
  #> Warning in mean.default(sort(x, partial = half + 0L:1L)[half + 0L:1L]):
  #> argument is not numeric or logical: returning NA
  #> Warning in mean.default(sort(x, partial = half + 0L:1L)[half + 0L:1L]):
  #> argument is not numeric or logical: returning NA
  #> Warning in mean.default(sort(x, partial = half + 0L:1L)[half + 0L:1L]):
  #> argument is not numeric or logical: returning NA
  #> Warning in mean.default(sort(x, partial = half + 0L:1L)[half + 0L:1L]):
  #> argument is not numeric or logical: returning NA
  #> Warning in mean.default(sort(x, partial = half + 0L:1L)[half + 0L:1L]):
  #> argument is not numeric or logical: returning NA
  ```
- [x] I used `covr` to analyze test coverage and got the following:
  ```
  > library(covr)
  > package_coverage()
  scToppR Coverage: 24.21%
  R/topp_plot.R: 0.00%
  R/toppFun.R: 38.01%
  R/zzz.R: 100.00%
  ```
  I recommend more tests for toppFun.R and topp_plot.R. For your save functions,
  make sure that you save to a temp directory when use is not interactive as
  we shouldn't automatically write to a user's home directory.
- [x] Add `BiocStyle` to Suggests in the DESCRIPTION file.
- [x] I recommend using `curl::has_internet()` to check for a connection for the functions that require an internet connection rather than using `eval=FALSE` or `dontrun` so that the functions can be run, which is better for catching bugs.
- [x] For toppSave where the example has `dontrun`, I recommend saving to a temporary directory so that it can be run as functions shouldn't write to the user's directory when run noninteractively.
- [x] Avoid using `cat` as a variable name.

**Comment by bryanwgranger** [comment]
Hello @jwokaty! Thank you for the feedback, I will address all of this in the next couple days.

**Comment by lshep** [comment]
@bryanwgranger  may we expect updates soon?

**Comment by bryanwgranger** [comment]
Yes, I have pushed a few smaller changes. I'm finishing everything else and will push those within a week.

**Comment by jwokaty** [comment]
Hi @bryanwgranger ,

After reviewing the package again, I realize that you may have not been finished updating. My apologies if that's the case. There are still a few remaining issues. To keep everything consolidated, I will update the [checklist](https://github.com/Bioconductor/Contributions/issues/3523#issuecomment-3352926660).

Although I didn't recommend it in my initial review, I recommend that for functions that require a connection that you check for it with `curl::has_internet()` so that these functions can be run in the vignettes. In the long run, it will help identify issues if these functions run. You'll need to add `curl` to `Suggests` in the `DESCRIPTION` file.

Thanks for your work and happy holidays.

**Comment by jwokaty** [comment]
Hi @bryanwgranger I'm following up to see if there will be additional updates?

**Comment by bryanwgranger** [comment]
Yes, I've been working on it and should finish updating by tomorrow, thanks!

**Comment by bryanwgranger** [comment]
Hey @jwokaty, thanks again for your feedback. I’ve gone through the checklist and addressed all the unchecked points:

- no disabled code blocks in vignette - enabled the saving/downloading blocks and `tempdir()` and used `curl::has_internet()` respectively. 

- coding and syntax - i’ve addressed the examples here and checked the contributors guide

- fixed the README installation instructions

- testing - I’ve added many more tests 

- removed warnings for toppBalloon

- use `curl::has_internet()` in vignette

- use `tempdir()` in all examples when saving instead of don’t run

- avoid using cat as variable name - renamed these variables

- I’ve also added seamless integration with SE/SCE objects - users can add toppdata results directly to an SE/SCE object (with function `sce <- addToppData(sce, toppData)` and plot directly (e.g. `toppPlot(sce, category = …)`

**Comment by lshep** [comment]
The default branch of your GitHub repository has been added to Bioconductor's
git repository as branch devel.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/bryanwgranger.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("scToppR")`. The package 'landing page' will be created at

https://bioconductor.org/packages/scToppR

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
